### PR TITLE
Various performance and stability improvements for NIO.

### PR DESCRIPTION
**Pull request by shorea**

### CODE REVIEW

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/MakeAsyncHttpRequestStage.java [hunk 1]
```diff
@@ -167,19 +182,25 @@ public void onStream(Publisher<ByteBuffer> publisher) {
 
         @Override
         public void exceptionOccurred(Throwable throwable) {
-            responseHandler.exceptionOccurred(throwable);
+            runAndLogError("SdkHttpResponseHandler threw an exception.",
```
Comment (shorea) [inline core/src/main/java/software/amazon/awssdk/http/pipeline/stages/MakeAsyncHttpRequestStage.java]: We weren't notifying the future in certain cases when the response handlers threw exceptions.

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/ResponseHandler.java [hunk 2]
```diff
@@ -55,61 +56,100 @@
...
         RequestContext requestContext = ctx.channel().attr(REQUEST_CONTEXT_KEY).get();
-        log.error("Exception processing request: {}", requestContext.sdkRequest(), cause);
-        requestContext.handler().exceptionOccurred(cause);
-        requestContext.channelPool().release(ctx.channel());
-        ctx.fireExceptionCaught(cause);
+        try {
+            log.error("Exception processing request: {}", requestContext.sdkRequest(), cause);
+            ctx.fireExceptionCaught(cause);
+        } finally {
+            runAndLogError("SdkHttpResponseHandler threw an exception",
```
Comment (shorea) [inline http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/ResponseHandler.java]: Makes sure we always release the channel back to the pool

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/RunnableRequest.java [hunk 3]
```diff
@@ -45,47 +55,58 @@ public RunnableRequest(RequestContext context) {
...
+    @Override
+    public void abort() {
+        if (channel != null) {
+            channel.disconnect().addListener(ignored -> context.channelPool().release(channel));
+        }
+    }
+
+    private void makeRequest(HttpRequest request) {
+        log.debug("Writing request: " + request);
+        channel.writeAndFlush(new StreamedRequest(context.nettyRequest(), context.sdkRequestProvider(), channel))
```
Comment (shorea) [inline http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/RunnableRequest.java]: writeAndFlush, at low TPS this was buffering to a certain amount before sending the request resulting in higher latencies.

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/SdkChannelPoolMap.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Replacement for {@link io.netty.channel.pool.AbstractChannelPoolMap}. This implementation guarentees
+ * only one instance of a {@link ChannelPool} is created for each key.
+ */
+public abstract class SdkChannelPoolMap<K, P extends ChannelPool>
```
Comment (shorea) [inline http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/SdkChannelPoolMap.java]: AbstractChannelPoolMap has some poor synchronization that results in creating unused channel maps.

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/RunnableRequest.java [hunk 5]
```diff
@@ -45,47 +55,58 @@ public RunnableRequest(RequestContext context) {
...
-        context.sdkRequestProvider().subscribe(subscriber);
+    @Override
+    public void abort() {
+        if (channel != null) {
+            channel.disconnect().addListener(ignored -> context.channelPool().release(channel));
+        }
+    }
+
+    private void makeRequest(HttpRequest request) {
+        log.debug("Writing request: " + request);
```
Comment (dagnir) [inline http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/RunnableRequest.java]: Should use "{}" placeholder

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/SdkChannelPoolMap.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.nio.netty.internal;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+
+import io.netty.channel.pool.ChannelPool;
+import io.netty.channel.pool.ChannelPoolMap;
+import io.netty.util.internal.ReadOnlyIterator;
```
Comment (dagnir) [inline http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/SdkChannelPoolMap.java]: Shouldn't use internal class per https://github.com/netty/netty/blob/566566db3a02de9f2170a6a42c95e69b4a628548/common/src/main/java/io/netty/util/internal/package-info.java#L18

Comment (shorea) [inline http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/SdkChannelPoolMap.java]: Good catch, copy-pasta'd this from netty. Will find an alternative.

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/RunnableRequest.java (hunk 5 above)
Comment (shorea) [inline http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/RunnableRequest.java]: +1

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/MakeAsyncHttpRequestStage.java [hunk 7]
```diff
@@ -167,19 +182,25 @@ public void onStream(Publisher<ByteBuffer> publisher) {
...
+                SdkHttpFullResponse httpFullResponse = (SdkHttpFullResponse) this.response;
+
+                publishProgress(listener, ProgressEventType.HTTP_REQUEST_COMPLETED_EVENT);
+                final HttpResponse httpResponse = SdkHttpResponseAdapter.adapt(false, request, httpFullResponse);
+                Response<OutputT> toReturn = handleResponse(httpResponse);
+                future.complete(toReturn);
+                return toReturn;
+            } catch (Exception e) {
+                future.completeExceptionally(e);
+                throw e;
```
Comment (spfink) [inline core/src/main/java/software/amazon/awssdk/http/pipeline/stages/MakeAsyncHttpRequestStage.java]: log the error?

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/SdkChannelPoolMap.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.nio.netty.internal;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
```
Comment (spfink) [inline http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/SdkChannelPoolMap.java]: Use our validators to keep messaging/logging consistent?

File: core/src/main/java/software/amazon/awssdk/http/pipeline/stages/MakeAsyncHttpRequestStage.java (hunk 7 above)
Comment (shorea) [inline core/src/main/java/software/amazon/awssdk/http/pipeline/stages/MakeAsyncHttpRequestStage.java]: We log it further up in RetryableStage at debug.

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/SdkChannelPoolMap.java (hunk 8 above)
Comment (shorea) [inline http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/SdkChannelPoolMap.java]: +1, done

**Pull request merged**